// @tanstack/form-core@1.33.0/dist/cjs/index.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const FormApi = require("./FormApi.cjs");
const FieldApi = require("./FieldApi.cjs");
const FormGroupApi = require("./FormGroupApi.cjs");
const utils = require("./utils.cjs");
const mergeForm = require("./mergeForm.cjs");
const formOptions = require("./formOptions.cjs");
const standardSchemaValidator = require("./standardSchemaValidator.cjs");
const FieldGroupApi = require("./FieldGroupApi.cjs");
const ValidationLogic = require("./ValidationLogic.cjs");
const EventClient = require("./EventClient.cjs");
const transform = require("./transform.cjs");
exports.FormApi = FormApi.FormApi;
exports.FieldApi = FieldApi.FieldApi;
exports.FormGroupApi = FormGroupApi.FormGroupApi;
exports.getDefaultFormGroupMeta = FormGroupApi.getDefaultFormGroupMeta;
exports.concatenatePaths = utils.concatenatePaths;
exports.createFieldMap = utils.createFieldMap;
exports.deepCopy = utils.deepCopy;
exports.deleteBy = utils.deleteBy;
exports.determineFieldLevelErrorSourceAndValue = utils.determineFieldLevelErrorSourceAndValue;
exports.determineFormLevelErrorSourceAndValue = utils.determineFormLevelErrorSourceAndValue;
exports.evaluate = utils.evaluate;
exports.functionalUpdate = utils.functionalUpdate;
exports.getAsyncValidatorArray = utils.getAsyncValidatorArray;
exports.getBy = utils.getBy;
exports.getSyncValidatorArray = utils.getSyncValidatorArray;
exports.isFieldInGroup = utils.isFieldInGroup;
exports.isGlobalFormValidationError = utils.isGlobalFormValidationError;
exports.isNonEmptyArray = utils.isNonEmptyArray;
exports.makePathArray = utils.makePathArray;
exports.mergeOpts = utils.mergeOpts;
exports.setBy = utils.setBy;
exports.throttleFormState = utils.throttleFormState;
exports.uuid = utils.uuid;
exports.mergeForm = mergeForm.mergeForm;
exports.mutateMergeDeep = mergeForm.mutateMergeDeep;
exports.formOptions = formOptions.formOptions;
exports.isStandardSchemaValidator = standardSchemaValidator.isStandardSchemaValidator;
exports.standardSchemaValidators = standardSchemaValidator.standardSchemaValidators;
exports.FieldGroupApi = FieldGroupApi.FieldGroupApi;
exports.defaultValidationLogic = ValidationLogic.defaultValidationLogic;
exports.revalidateLogic = ValidationLogic.revalidateLogic;
exports.formEventClient = EventClient.formEventClient;
exports.mergeAndUpdate = transform.mergeAndUpdate;
